### PR TITLE
[SYCL] Share PFWG lambda object through shared memory

### PR DESCRIPTION
**Pull request by againull**
In the current implementation private address of the PFWG lambda object
is shared by leader work item through local memory to other work items.
This is not correct. That is why perform the copy of the PFWG lambda
object to shared memory and make work items work with address of the
object in shared memory. I.e. this case should be handled in the
similar way as for byval parameters.

Signed-off-by: Artur Gainullin <artur.gainullin@intel.com>

### CODE REVIEW

File: llvm/test/SYCLLowerIR/pfwg_and_pfwi.ll [hunk 1]
```diff
@@ -13,51 +13,55 @@
...
 ; CHECK-NEXT:  bb:
+; CHECK-NEXT:    [[TMP:%.*]] = alloca [[STRUCT_BAR:%.*]] addrspace(4)*, align 8
+; CHECK-NEXT:    [[TMP2:%.*]] = alloca [[STRUCT_FOO_0:%.*]], align 1
 ; CHECK-NEXT:    [[TMP0:%.*]] = load i64, i64 addrspace(1)* @__spirv_BuiltInLocalInvocationIndex
 ; CHECK-NEXT:    [[CMPZ3:%.*]] = icmp eq i64 [[TMP0]], 0
 ; CHECK-NEXT:    br i1 [[CMPZ3]], label [[LEADER:%.*]], label [[MERGE:%.*]]
 ; CHECK:       leader:
 ; CHECK-NEXT:    [[TMP1:%.*]] = bitcast %struct.zot* [[ARG1:%.*]] to i8*
-; CHECK-NEXT:    call void @llvm.memcpy.p3i8.p0i8.i64(i8 addrspace(3)* align 16 bitcast (%struct.zot addrspace(3)* @[[GROUP_SHADOW]] to i8 addrspace(3)*), i8* align 8 [[TMP1]], i64 96, i1 false)
+; CHECK-NEXT:    call void @llvm.memcpy.p3i8.p0i8.i64(i8 addrspace(3)* align 16 bitcast (%struct.zot addrspace(3)* @ArgShadow.2 to i8 addrspace(3)*), i8* align 8 [[TMP1]], i64 96, i1 false)
```
Comment (kbobrovs) [inline llvm/test/SYCLLowerIR/pfwg_and_pfwi.ll:32]: Please consider replacing `ArgShadow.2` with a match (added above at the declaration point)

File: llvm/lib/SYCLLowerIR/LowerWGScope.cpp [hunk 2]
```diff
@@ -679,10 +688,19 @@ static void fixupPrivateMemoryPFWILambdaCaptures(CallInst *PFWICall) {
 // Go through "byval" parameters which are passed as AS(0) pointers
 // and: (1) create local shadows for them (2) and initialize them from the
 // leader's copy and (3) replace usages with pointer to the shadow
+//
+// Do the same with PFWG lamda object parameter ('this' pointer) which is also
+// passed as AS(0) pointer.
```
[the remark this answers is not in the document]
Comment (againull) [reply llvm/lib/SYCLLowerIR/LowerWGScope.cpp]: Thx, added the comment why copy back is not needed.

File: llvm/lib/SYCLLowerIR/LowerWGScope.cpp [hunk 3]
```diff
@@ -679,10 +688,19 @@ static void fixupPrivateMemoryPFWILambdaCaptures(CallInst *PFWICall) {
...
 static void shareByValParams(Function &F, const Triple &TT) {
```
[the remark this answers is not in the document]
Comment (againull) [reply llvm/lib/SYCLLowerIR/LowerWGScope.cpp]: Fixed

File: llvm/lib/SYCLLowerIR/LowerWGScope.cpp [hunk 4]
```diff
@@ -692,38 +710,46 @@ static void shareByValParams(Function &F, const Triple &TT) {
...
+        // LLVM default address space 0 which is the generic address space for
+        // NVPTX target.
+        assert(Arg.getType()->getPointerAddressSpace() == 0);
+
+        // Cast a pointer in the shared address space to the generic address
+        // space.
+        RepVal = ConstantExpr::getPointerBitCastOrAddrSpaceCast(Shadow,
+                                                                Arg.getType());
+      }
+    } else if (Arg.getArgNo() == 0) {
```
[the remark this answers is not in the document]
Comment (againull) [reply llvm/lib/SYCLLowerIR/LowerWGScope.cpp]: Fixed

File: llvm/test/SYCLLowerIR/pfwg_and_pfwi.ll (hunk 1 above)
Comment (againull) [reply llvm/test/SYCLLowerIR/pfwg_and_pfwi.ll]: Fixed

**Pull request merged**